dpctl/tensor/_dlpack.pyx

```diff
@@ -1005,31 +1005,60 @@ def from_dlpack(x, /, *, device=None, copy=None):
             X = dpt.from_dlpack(C)
 
     """
-    if not hasattr(x, "__dlpack__"):
+    dlpack_attr = getattr(x, "__dlpack__", None)
+    dlpack_dev_attr = getattr(x, "__dlpack_device__", None)
+    if not callable(dlpack_attr) or not callable(dlpack_dev_attr):
         raise TypeError(
             f"The argument of type {type(x)} does not implement "
-            "`__dlpack__` method."
-        )
-    dlpack_attr = getattr(x, "__dlpack__")
-    if not callable(dlpack_attr):
-        raise TypeError(
-            f"The argument of type {type(x)} does not implement "
-            "`__dlpack__` method."
+            "`__dlpack__` and `__dlpack_device__` methods."
         )
     try:
         # device is converted to a dlpack_device if necessary
         dl_device = None
         if device:
             if isinstance(device, tuple):
                 dl_device = device
+                if len(dl_device) != 2:
+                    raise ValueError(
+                        "Argument `device` specified as a tuple must have length 2"
+                    )
             else:
                 if not isinstance(device, dpctl.SyclDevice):
                     d = Device.create_device(device).sycl_device
-                    dl_device = (device_OneAPI, get_parent_device_ordinal_id(<c_dpctl.SyclDevice>d))
                 else:
-                    dl_device = (device_OneAPI, get_parent_device_ordinal_id(<c_dpctl.SyclDevice>device))
+                    d = device
+                dl_device = (device_OneAPI, get_parent_device_ordinal_id(<c_dpctl.SyclDevice>d))
         dlpack_capsule = dlpack_attr(max_version=get_build_dlpack_version(), dl_device=dl_device, copy=copy)
         return from_dlpack_capsule(dlpack_capsule)
     except TypeError:
-        dlpack_capsule = dlpack_attr()
-        return from_dlpack_capsule(dlpack_capsule)
+        x_dldev = dlpack_dev_attr()
+        if (dl_device is None) or (dl_device == x_dldev):
+            dlpack_capsule = dlpack_attr()
+            return from_dlpack_capsule(dlpack_capsule)
+        # must copy via host
+        if copy is False:
+            raise ValueError(
```

Comment:
Array API recommends this case to be a `BufferError` for `__dlpack__`, I think that we should be consistent here.